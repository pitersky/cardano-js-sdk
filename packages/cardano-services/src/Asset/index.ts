export * from './CardanoTokenRegistry';
export * from './DbSyncNftMetadataService';
export * from './types';
