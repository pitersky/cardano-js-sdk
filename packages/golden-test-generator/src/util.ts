import { getLastCommit } from 'git-last-commit';
// eslint-disable-next-line unicorn/prefer-node-protocol
import { promisify } from 'util';

export const getLastCommitPromise = promisify(getLastCommit);
