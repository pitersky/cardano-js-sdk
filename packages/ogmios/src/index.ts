export * from './TxSubmitProvider';
export * from './TimeSettingsProvider';
export * from './util';
export { createConnectionObject } from '@cardano-ogmios/client';
