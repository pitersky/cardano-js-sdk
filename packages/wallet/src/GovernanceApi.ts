import { AsyncKeyAgent, util as keyManagementUtil } from './KeyManagement';
import { CSL, Cardano, util } from '@cardano-sdk/core';

export type BuildDelegationProps = {
  votingPublicKey: Cardano.Bip32PublicKey;
  nonce: number;
  rewardAccount: Cardano.RewardAccount;
};

export interface SignDelegationProps {
  votingPublicKey: string;
  publicStakeKey: string;
  rewardAccountKey: string;
  nonce: number;
}

export interface VotingKeyPair {
  prvKey: Cardano.Bip32PrivateKey;
  pubKey: Cardano.Bip32PublicKey;
}

export enum VotingLabels {
  DATA = 61_284,
  SIG = 61_285
}

/**
 * TODO: Maybe it would be good to treat this as a BASE and create additional, key agent dependent, GOV API's
 */
export class GovernanceApi {
  readonly #keyAgent: AsyncKeyAgent;

  constructor(keyAgent: AsyncKeyAgent) {
    this.#keyAgent = keyAgent;
  }

  getVotingKey(): VotingKeyPair {
    const mnemonic = keyManagementUtil.generateMnemonicWords(160);
    const bip39entropy = keyManagementUtil.mnemonicWordsToEntropy(mnemonic);
    const EMPTY_PASSWORD = Buffer.from('');
    const cslPrvKey = CSL.Bip32PrivateKey.from_bip39_entropy(Buffer.from(bip39entropy, 'hex'), EMPTY_PASSWORD);
    const cslPrvKeyBytes = Buffer.from(cslPrvKey.as_bytes()).toString('hex');
    const cslPubKeyBytes = Buffer.from(cslPrvKey.to_public().as_bytes()).toString('hex');

    return {
      prvKey: Cardano.Bip32PrivateKey(cslPrvKeyBytes),
      pubKey: Cardano.Bip32PublicKey(cslPubKeyBytes)
    };
  }

  async buildDelegation({
    votingPublicKey,
    nonce,
    rewardAccount
  }: BuildDelegationProps): Promise<Cardano.AuxiliaryData> {
    const publicStakeKey = await this.#keyAgent.derivePublicKey(keyManagementUtil.STAKE_KEY_DERIVATION_PATH);
    const cslRewardAccount = CSL.Address.from_bech32(rewardAccount.toString());
    const rewardAccountBytes = Buffer.from(cslRewardAccount.to_bytes());
    const votingData = new Map<bigint, Cardano.Metadatum>([
      [1n, Buffer.from(votingPublicKey, 'hex')],
      [2n, Buffer.from(publicStakeKey, 'hex')],
      [3n, rewardAccountBytes],
      [4n, BigInt(nonce)]
    ]);
    const signature = await this.signDelegation({
      nonce,
      publicStakeKey: `0x${publicStakeKey}`,
      rewardAccountKey: `0x${rewardAccountBytes.toString('hex')}`,
      votingPublicKey: `0x${votingPublicKey}`
    });

    // blake2b-256 hash of the metadata signed using stakekey
    const votingSignature = new Map([[1n, Buffer.from(signature, 'hex')]]);

    const votingAuxData = new Map<bigint, Cardano.MetadatumMap>([
      [BigInt(VotingLabels.DATA), votingData],
      [BigInt(VotingLabels.SIG), votingSignature]
    ]);

    return {
      body: {
        blob: votingAuxData
      }
    };
  }

  async signDelegation({
    votingPublicKey,
    publicStakeKey,
    rewardAccountKey,
    nonce
  }: SignDelegationProps): Promise<Cardano.Ed25519Signature> {
    const blake2b = require('blake2b');
    const { index, role } = keyManagementUtil.STAKE_KEY_DERIVATION_PATH;
    const generalMetadata = CSL.GeneralTransactionMetadata.new();
    const registrationData = CSL.encode_json_str_to_metadatum(
      JSON.stringify({
        '1': votingPublicKey,
        '2': publicStakeKey,
        '3': rewardAccountKey,
        '4': nonce
      }),
      CSL.MetadataJsonSchema.BasicConversions
    );
    generalMetadata.insert(CSL.BigNum.from_str(VotingLabels.DATA.toString()), registrationData);
    const hashedMetadata = blake2b(256 / 8)
      .update(generalMetadata.to_bytes())
      .digest('binary');

    const { signature } = await this.#keyAgent.signBlob({ index, role }, util.bytesToHex(hashedMetadata));
    return signature;
  }
}
