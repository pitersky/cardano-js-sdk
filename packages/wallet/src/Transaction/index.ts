export * from './createTransactionInternals';
export * from './ensureValidityInterval';
