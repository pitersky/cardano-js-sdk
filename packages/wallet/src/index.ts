export * as Transaction from './Transaction';
export * as KeyManagement from './KeyManagement';
export * from './SingleAddressWallet';
export * from './types';
export * from './services';
export * as storage from './persistence';
export * as cip30 from './cip30';
export * from './GovernanceApi';
