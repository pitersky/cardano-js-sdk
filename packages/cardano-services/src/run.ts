#!/usr/bin/env node
import * as envalid from 'envalid';
import { API_URL_DEFAULT, OGMIOS_URL_DEFAULT, RABBITMQ_URL_DEFAULT, ServiceNames, loadHttpServer } from './Program';
import { CACHE_TTL_DEFAULT } from './InMemoryCache';
import { LogLevel } from 'bunyan';
import { POLL_INTERVAL_DEFAULT } from './NetworkInfo';
import { URL } from 'url';
import { USE_QUEUE_DEFAULT } from './ProgramsCommon';
import { cacheTtlValidator } from './util/validators';
import { config } from 'dotenv';
import { loggerMethodNames } from './util';
import onDeath from 'death';

const envSpecs = {
  API_URL: envalid.url({ default: API_URL_DEFAULT }),
  CACHE_TTL: envalid.makeValidator(cacheTtlValidator)(envalid.num({ default: CACHE_TTL_DEFAULT })),
  CARDANO_NODE_CONFIG_PATH: envalid.str({ default: undefined }),
  DB_CONNECTION_STRING: envalid.str({ default: undefined }),
  LOGGER_MIN_SEVERITY: envalid.str({ choices: loggerMethodNames as string[], default: 'info' }),
  OGMIOS_URL: envalid.url({ default: OGMIOS_URL_DEFAULT }),
  POLL_INTERVAL: envalid.num({ default: POLL_INTERVAL_DEFAULT }),
  RABBITMQ_URL: envalid.url({ default: RABBITMQ_URL_DEFAULT }),
  SERVICE_NAMES: envalid.str({ example: Object.values(ServiceNames).toString() }),
  USE_QUEUE: envalid.bool({ default: USE_QUEUE_DEFAULT })
};

void (async () => {
  config();
  const env = envalid.cleanEnv(process.env, envSpecs);
  const apiUrl = new URL(env.API_URL);
  const ogmiosUrl = new URL(env.OGMIOS_URL);
  const rabbitmqUrl = new URL(env.RABBITMQ_URL);
  const cardanoNodeConfigPath = env.CARDANO_NODE_CONFIG_PATH;
  const cacheTtl = env.CACHE_TTL;
  const pollInterval = env.POLL_INTERVAL;
  const dbConnectionString = env.DB_CONNECTION_STRING ? new URL(env.DB_CONNECTION_STRING).toString() : undefined;
  const serviceNames = env.SERVICE_NAMES.split(',') as ServiceNames[];

  try {
    const server = await loadHttpServer({
      apiUrl,
      options: {
        cacheTtl,
        cardanoNodeConfigPath,
        dbConnectionString,
        loggerMinSeverity: env.LOGGER_MIN_SEVERITY as LogLevel,
        ogmiosUrl,
        pollInterval,
        rabbitmqUrl,
        useQueue: env.USE_QUEUE
      },
      serviceNames
    });
    await server.initialize();
    await server.start();
    onDeath(async () => {
      await server.shutdown();
      // eslint-disable-next-line unicorn/no-process-exit
      process.exit(1);
    });
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(1);
  }
})();
