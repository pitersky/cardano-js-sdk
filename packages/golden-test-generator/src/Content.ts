import { Commit } from 'git-last-commit';
import { Ogmios } from '@cardano-sdk/ogmios';
import { getLastCommitPromise } from './util';
const packageJson = require('../package.json');

export type Metadata = {
  cardano: {
    compactGenesis: Ogmios.Schema.CompactGenesis;
    intersection: Ogmios.ChainSync.Intersection;
  };
  software: {
    name: string;
    version: string;
    commit: Pick<Commit, 'hash' | 'tags'>;
  };
};

export type GeneratorMetadata = { metadata: { cardano: Metadata['cardano'] } };

export const prepareContent = async <Body>(
  metadata: Omit<Metadata, 'software'>,
  body: Body
): Promise<{
  metadata: Metadata;
  body: Body;
}> => {
  const lastCommit = await getLastCommitPromise();
  return {
    body,
    metadata: {
      ...metadata,

      software: {
        commit: {
          hash: lastCommit.hash,
          tags: lastCommit.tags
        },
        name: packageJson.name,
        version: packageJson.version
      }
    }
  };
};
