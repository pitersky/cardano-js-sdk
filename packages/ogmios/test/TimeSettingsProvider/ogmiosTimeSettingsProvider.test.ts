import { Connection, createConnectionObject } from '@cardano-ogmios/client';
import { ProviderError, TimeSettingsProvider } from '@cardano-sdk/core';
import { createMockOgmiosServer, listenPromise, serverClosePromise } from '../mocks/mockOgmiosServer';
import { getRandomPort } from 'get-port-please';
import { ogmiosTimeSettingsProvider } from '../../src';
import http from 'http';

describe('ogmiosTimeSettingsProvider', () => {
  let mockServer: http.Server;
  let connection: Connection;
  let provider: TimeSettingsProvider;

  beforeAll(async () => {
    connection = createConnectionObject({ port: await getRandomPort() });
  });
  describe('healthCheck', () => {
    afterEach(async () => {
      if (mockServer !== undefined) {
        await serverClosePromise(mockServer);
      }
    });

    it('is not ok if cannot connect', async () => {
      provider = ogmiosTimeSettingsProvider(connection);
      const res = await provider.healthCheck();
      expect(res).toEqual({ ok: false });
    });

    it('is ok if node is close to the network tip', async () => {
      mockServer = createMockOgmiosServer({
        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
      });
      await listenPromise(mockServer, connection.port);
      provider = ogmiosTimeSettingsProvider(connection);
      const res = await provider.healthCheck();
      expect(res).toEqual({ ok: true });
    });

    it('is not ok if node is not close to the network tip', async () => {
      mockServer = createMockOgmiosServer({
        healthCheck: { response: { networkSynchronization: 0.8, success: true } }
      });
      await listenPromise(mockServer, connection.port);
      provider = ogmiosTimeSettingsProvider(connection);
      const res = await provider.healthCheck();
      expect(res).toEqual({ ok: false });
    });

    it('throws a typed error if caught during the service interaction', async () => {
      mockServer = createMockOgmiosServer({
        healthCheck: { response: { failWith: new Error('Some error'), success: false } }
      });
      await listenPromise(mockServer, connection.port);
      provider = ogmiosTimeSettingsProvider(connection);
      await expect(provider.healthCheck()).rejects.toThrowError(ProviderError);
    });
  });

  describe('timeSettings', () => {
    describe('success', () => {
      beforeAll(async () => {
        mockServer = createMockOgmiosServer({ timeSettings: { response: { success: true } } });
        await listenPromise(mockServer, connection.port);
        provider = ogmiosTimeSettingsProvider(connection);
      });

      afterAll(async () => {
        await serverClosePromise(mockServer);
      });

      it('resolves if successful', async () => {
        const res = await provider.timeSettings();
        expect(res).toMatchSnapshot();
      });
    });

    describe('failure', () => {
      beforeAll(async () => {
        mockServer = createMockOgmiosServer({
          timeSettings: { response: { failWith: { type: 'queryUnavailableInEra' }, success: false } }
        });
        await listenPromise(mockServer, connection.port);
        provider = ogmiosTimeSettingsProvider(connection);
      });
      afterEach(async () => {
        await serverClosePromise(mockServer);
      });

      it('rejects with errors thrown by the service', async () => {
        await expect(provider.timeSettings()).rejects.toThrowError(ProviderError);
      });
    });
  });
});
