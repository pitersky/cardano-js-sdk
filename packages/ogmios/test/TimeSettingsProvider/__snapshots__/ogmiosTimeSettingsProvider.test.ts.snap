// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ogmiosTimeSettingsProvider timeSettings success resolves if successful 1`] = `
Array [
  Object {
    "epochLength": 21600,
    "fromSlotDate": 2017-09-23T21:44:51.000Z,
    "fromSlotNo": 0,
    "slotLength": 20,
  },
  Object {
    "epochLength": 432000,
    "fromSlotDate": 2018-09-28T21:44:51.000Z,
    "fromSlotNo": 1598400,
    "slotLength": 1,
  },
]
`;
